### PR TITLE
Changes to PGen for better user experience

### PR DESCRIPTION
**Pull request by hamidhandid**
Hi.
For better user experience I added 4 feature to PGen. As a user when I wanted to use the PGen, I had some problems with it. I think this features can help users.

**Features:**

1. I add apply with enter key when user wants to edit an edge properties. This means user can press enter key to apply changes of edge like pressing apply button. This change is applied in EdgePropertiesDialog.fxml & EdgePropertiesController.java files.
2. I change file chooser and directory chooser to singleton ones. This means when user saves or opens a file once, in other savings or openings of files the last path that user had chosen, can be remembered. For this change I add SingletonChooser.java and change DialogUtility.java.
3. When user press Check Graphs, if an edge has no token, in PGen the error was shown only with "Token can not be empty" message. I add better error with giving start and end node of the edge that has no token like "Token can not be empty between start node 0 and end node 1". This change is applied in CheckUtility.java & LLParserGenerator.java.
4. Sometimes users may forget save files and press close button of program accidentally. For this problem, I add setOnCloseRequest to warn user whether he/she wants to save the program or not. This change is applied in Main.java & MainController.java.

I think merging this changes can give better experiences to users when they use PGen.
Thanks.

### CODE REVIEW

File: src/main/java/ir/ac/sbu/controller/EdgePropertiesController.java [hunk 1]
```diff
@@ -35,6 +37,16 @@ public void init(EdgeModel edge) {
...
+        applyFunc();
+    }
+
+    public void applyWithEnterKey(KeyEvent keyEvent) {
+        if (keyEvent.getCode().equals(KeyCode.ENTER)) {
+            applyFunc();
+        }
+    }
+
+    private void applyFunc() {
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/controller/EdgePropertiesController.java:49]: Rename function name to "applyChanges"

File: src/main/java/ir/ac/sbu/controller/EdgePropertiesController.java [hunk 2]
```diff
@@ -35,6 +37,16 @@ public void init(EdgeModel edge) {
     }
 
     public void apply(ActionEvent actionEvent) {
+        applyFunc();
+    }
+
+    public void applyWithEnterKey(KeyEvent keyEvent) {
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/controller/EdgePropertiesController.java]: `public void handleKeyPressed(KeyEvent keyEvent) {`

File: src/main/java/ir/ac/sbu/controller/EdgePropertiesController.java (hunk 1 above)
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/controller/EdgePropertiesController.java]: `private void applyChanges() {`

File: src/main/resources/fxml/EdgePropertiesDialog.fxml [hunk 3]
```diff
@@ -25,20 +25,20 @@
         <Insets bottom="10.0" left="10.0" right="10.0" top="10.0" />
     </padding>
     <Label text="Token" />
-    <TextField fx:id="tokenText" GridPane.columnIndex="1" GridPane.hgrow="ALWAYS" GridPane.vgrow="ALWAYS">
+    <TextField fx:id="tokenText" GridPane.columnIndex="1" GridPane.hgrow="ALWAYS" GridPane.vgrow="ALWAYS" onKeyPressed="#applyWithEnterKey">
```
Comment (Borjianamin98) [inline src/main/resources/fxml/EdgePropertiesDialog.fxml]: Instead of providing this functionality for every button and element in page (which is not well developed and on the other hand whenever a case is added, one needs to pay attention to this), provide it as a general handler for whole page:
`<GridPane xmlns="http://javafx.com/javafx/8.0.171" xmlns:fx="http://javafx.com/fxml/1"
  fx:controller="ir.ac.sbu.controller.EdgePropertiesController" onKeyPressed="#handleKeyPressed">`

File: src/main/java/ir/ac/sbu/utility/SingletonChooser.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
+package ir.ac.sbu.utility;
+
+import javafx.beans.property.SimpleObjectProperty;
+import javafx.stage.DirectoryChooser;
+import javafx.stage.FileChooser;
+import javafx.stage.Window;
+
+import java.io.File;
+
+class SingletonDirectoryChooser {
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/SingletonChooser.java]: Instead of using non public class, please add a directory with name "chooser" and provide this two class as separate files in that folder.

File: src/main/java/ir/ac/sbu/utility/SingletonChooser.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
+package ir.ac.sbu.utility;
+
+import javafx.beans.property.SimpleObjectProperty;
+import javafx.stage.DirectoryChooser;
+import javafx.stage.FileChooser;
+import javafx.stage.Window;
+
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/SingletonChooser.java]: Remove empty lines between imports. I will provide automatic checkstyle for this project later.

File: src/main/java/ir/ac/sbu/utility/DialogUtility.java [hunk 6]
```diff
@@ -37,23 +37,20 @@ private DialogUtility() {
...
+        SingletonFileChooser.setTitle("Open a File");
+        return SingletonFileChooser.showOpenDialog(owner);
     }
 
     public static File showDirectoryDialog(Window owner) {
-        DirectoryChooser chooser = new DirectoryChooser();
-        chooser.setTitle("Select a directory");
-        return chooser.showDialog(owner);
+        SingletonDirectoryChooser.setTitle("Select a directory");
+        return SingletonDirectoryChooser.showDialog(owner);
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/DialogUtility.java]: Do this after reading my other comments.

```
return SingletonDirectoryChooser.getInstance()
                .setTitle("Select a directory")
                .showDialog(owner);
```

File: src/main/java/ir/ac/sbu/utility/SingletonChooser.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import java.io.File;
+
+class SingletonDirectoryChooser {
+    private static DirectoryChooser instance = null;
+    private static SimpleObjectProperty<File> lastKnownDirectoryProperty = new SimpleObjectProperty<>();
+
+    private SingletonDirectoryChooser() {
+    }
+
+    private static DirectoryChooser getInstance() {
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/SingletonChooser.java]: Use below better approach for singleton classes:
```
class SingletonDirectoryChooser {

    private final static SingletonDirectoryChooser INSTANCE = new SingletonDirectoryChooser();

    private final DirectoryChooser directoryChooser;
    private final SimpleObjectProperty<File> lastKnownDirectoryProperty;

    private SingletonDirectoryChooser() {
        this.directoryChooser = new DirectoryChooser();
        this.lastKnownDirectoryProperty = new SimpleObjectProperty<>();
        this.directoryChooser.initialDirectoryProperty().bindBidirectional(lastKnownDirectoryProperty);
    }

    public static SingletonDirectoryChooser getInstance() {
        return INSTANCE;
    }

    public SingletonDirectoryChooser setTitle(String title) {
        directoryChooser.setTitle(title);
        return this;
    }

    public File showDialog(Window ownerWindow) {
        File chosenDirectory = directoryChooser.showDialog(ownerWindow);
        if (chosenDirectory != null) {
            lastKnownDirectoryProperty.setValue(chosenDirectory.getAbsoluteFile());
        }
        return chosenDirectory;
    }
}
```

File: src/main/java/ir/ac/sbu/utility/DialogUtility.java [hunk 8]
```diff
@@ -37,23 +37,20 @@ private DialogUtility() {
     }
 
     public static File showSaveDialog(Window owner, String title, String... extensions) {
-        FileChooser chooser = new FileChooser();
-        chooser.setTitle(title);
-        chooser.getExtensionFilters().addAll(new FileChooser.ExtensionFilter("extensions", extensions));
-        return chooser.showSaveDialog(owner);
+        SingletonFileChooser.setExtensions(extensions);
+        SingletonFileChooser.setTitle(title);
+        return SingletonFileChooser.showSaveDialog(owner);
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/DialogUtility.java]: Do this after read other comments:
```
return SingletonFileChooser.getInstance()
                .setExtensions(extensions)
                .setTitle(title)
                .showSaveDialog(owner);
```

File: src/main/java/ir/ac/sbu/utility/DialogUtility.java [hunk 9]
```diff
@@ -37,23 +37,20 @@ private DialogUtility() {
...
     }
 
     public static File showOpenDialog(Window owner, String... extensions) {
-        FileChooser chooser = new FileChooser();
-        chooser.setTitle("Open a File");
-        chooser.getExtensionFilters().addAll(new FileChooser.ExtensionFilter("extensions", extensions));
-        return chooser.showOpenDialog(owner);
+        SingletonFileChooser.setExtensions(extensions);
+        SingletonFileChooser.setTitle("Open a File");
+        return SingletonFileChooser.showOpenDialog(owner);
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/DialogUtility.java]: Do this after read other comments:
```
return SingletonFileChooser.getInstance()
                .setExtensions(extensions)
                .setTitle("Open a File")
                .showOpenDialog(owner);
```

File: src/main/java/ir/ac/sbu/utility/SingletonChooser.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+class SingletonFileChooser {
+    private static FileChooser instance = null;
+    private static SimpleObjectProperty<File> lastKnownDirectoryProperty = new SimpleObjectProperty<>();
+
+    private SingletonFileChooser() {
+    }
+
+    private static FileChooser getInstance() {
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/SingletonChooser.java]: Use bellow better approach:
```
package ir.ac.sbu.utility.chooser;

import java.io.File;
import javafx.beans.property.SimpleObjectProperty;
import javafx.stage.FileChooser;
import javafx.stage.Window;

public class SingletonFileChooser {

    private final static SingletonFileChooser INSTANCE = new SingletonFileChooser();

    private final FileChooser fileChooser;
    private final SimpleObjectProperty<File> lastKnownDirectoryProperty;

    private SingletonFileChooser() {
        this.fileChooser = new FileChooser();
        this.lastKnownDirectoryProperty = new SimpleObjectProperty<>();
        this.fileChooser.initialDirectoryProperty().bindBidirectional(lastKnownDirectoryProperty);
    }

    public static SingletonFileChooser getInstance() {
        return INSTANCE;
    }

    public SingletonFileChooser setExtensions(String... extensions) {
        fileChooser.getExtensionFilters().setAll(new FileChooser.ExtensionFilter("extensions", extensions));
        return this;
    }

    public SingletonFileChooser setTitle(String title) {
        fileChooser.setTitle(title);
        return this;
    }

    public File showOpenDialog(Window ownerWindow) {
        File chosenFile = fileChooser.showOpenDialog(ownerWindow);
        if (chosenFile != null) {
            lastKnownDirectoryProperty.setValue(chosenFile.getParentFile());
        }
        return chosenFile;
    }

    public File showSaveDialog(Window ownerWindow) {
        File chosenFile = fileChooser.showSaveDialog(ownerWindow);
        if (chosenFile != null) {
            lastKnownDirectoryProperty.setValue(chosenFile.getParentFile());
        }
        return chosenFile;
    }
}
```

File: src/main/java/ir/ac/sbu/utility/CheckUtility.java [hunk 11]
```diff
@@ -26,6 +26,19 @@ public static void checkTokenName(String tokenName) {
         }
     }
 
+    public static void checkEmptyTokenName(String tokenName, int startNodeId, int endNodeId) {
+        String tokenCanNotBeEmptyStr = "Token can not be empty with start node " + startNodeId + " and end node " + endNodeId;
+        if (tokenName == null) {
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/CheckUtility.java]: It couldn't be null so redundant check.

File: src/main/java/ir/ac/sbu/utility/CheckUtility.java [hunk 12]
```diff
@@ -26,6 +26,19 @@ public static void checkTokenName(String tokenName) {
         }
     }
 
+    public static void checkEmptyTokenName(String tokenName, int startNodeId, int endNodeId) {
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/CheckUtility.java]: All of thing you do here is same as what exists before: `checkTokenName` method of class.
Remove this and do what I suggested at above.

File: src/main/java/ir/ac/sbu/parser/LLParserGenerator.java [hunk 13]
```diff
@@ -97,17 +97,22 @@ private void checkEdges() throws TableException {
...
         List<String> messages = new ArrayList<>();
+        for (GraphModel graphModel : graphs) {
+            for (EdgeModel edgeModel : graphModel.getEdges()) {
+                try {
+                    CheckUtility.checkEmptyTokenName(edgeModel.getToken(), edgeModel.getStart().getId(), edgeModel.getEnd().getId());
+                } catch (IllegalArgumentException e) {
+                    messages.add(e.getMessage());
+                }
+            }
+        }
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/parser/LLParserGenerator.java]: Too much redundant code. Please do this:
1) First remove lines 43-57 of this file.
2) Change this method name to "extractTokens" and call it instead of removed lines in first step.
3) Do this in method:
```
List<String> messages = new ArrayList<>();

        List<EdgeModel> edges = graphs.stream()
                .flatMap(graph -> graph.getEdges().stream())
                .collect(Collectors.toList());

        variables = graphs.stream().map(GraphModel::getName).collect(Collectors.toSet());
        tokenAsInt = new HashMap<>();
        tokenAsInt.put("$", eofTokenId);
        int tokenUID = 1;
        for (EdgeModel edge : edges) {
            String token = edge.getToken();
            if (edge.isGraph()) {
                variables.add(token);
            } else {
                try {
                    CheckUtility.checkTokenName(token);
                } catch (IllegalArgumentException e) {
                    messages.add(String.format("Invalid token: start-node = %s end-node = %s error = %s",
                            edge.getStart().getId(), edge.getEnd().getId(), e.getMessage()));
                }
                tokenAsInt.put(token, tokenUID);
                tokenUID++;
            }
        }
```
4) Remove unnecessary codes from method after adding above code.

File: src/main/java/ir/ac/sbu/utility/CheckUtility.java [hunk 14]
```diff
@@ -26,6 +26,19 @@ public static void checkTokenName(String tokenName) {
...
+        String tokenCanNotBeEmptyStr = "Token can not be empty with start node " + startNodeId + " and end node " + endNodeId;
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/utility/CheckUtility.java]: You check different conditions but you say it is empty in final error. It is not good. I fix this in my suggestion.

File: src/main/java/ir/ac/sbu/controller/MainController.java [hunk 15]
```diff
@@ -358,18 +367,30 @@ private void fileOpen(ActionEvent actionEvent) {
...
 
     private void fileSave(ActionEvent actionEvent) {
+        _fileSave();
+    }
+
+    private void fileSaveAs(ActionEvent actionEvent) {
+        _fileSaveAs();
+    }
+
+    private void _fileSave() {
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/controller/MainController.java]: Revert this changes. See other comments.

File: src/main/java/ir/ac/sbu/controller/MainController.java [hunk 16]
```diff
@@ -419,4 +440,27 @@ private void showModal(URL resource, String title) {
...
+        Main.getPrimaryStage().setOnCloseRequest(event -> {
+            windowEvent = event;
+            String dialogString = lastChosenFileForSave != null ? "Do you want to save changes?" : "Do you want to save changes to a file?";
+            Optional<ButtonType> result = DialogUtility.showConfirmationDialog("PGEN", dialogString);
+            if (!result.isPresent()) {
+                return;
+            }
+            ButtonType selectedButton = result.get();
+            if (selectedButton.equals(ButtonType.YES)) {
+                _fileSave();
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/controller/MainController.java]: Call `fileSave(null)` instead of creating new functions.

File: src/main/java/ir/ac/sbu/controller/MainController.java [hunk 17]
```diff
@@ -358,18 +367,30 @@ private void fileOpen(ActionEvent actionEvent) {
...
-    private void fileSaveAs(ActionEvent actionEvent) {
+    private void _fileSaveAs() {
         File selectedFile = DialogUtility.showSaveDialog(pane.getScene().getWindow(), "Save PGen File", "*.pgs");
         if (selectedFile != null) {
             lastChosenFileForSave = selectedFile;
             saveGraphs(lastChosenFileForSave);
+        } else {
+            if (windowEvent != null) {
+                windowEvent.consume();
+            }
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/controller/MainController.java]: Base on lines 456-464, we always consume `windowEvent' so it should be redundant. Am I wrong? If not, remove it.

File: src/main/java/ir/ac/sbu/controller/MainController.java [hunk 18]
```diff
@@ -419,4 +440,27 @@ private void showModal(URL resource, String title) {
...
         }
     }
+
+    private void setOnCloseRequest() {
+        Main.getPrimaryStage().setOnCloseRequest(event -> {
+            windowEvent = event;
+            String dialogString = lastChosenFileForSave != null ? "Do you want to save changes?" : "Do you want to save changes to a file?";
+            Optional<ButtonType> result = DialogUtility.showConfirmationDialog("PGEN", dialogString);
+            if (!result.isPresent()) {
+                return;
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/controller/MainController.java]: It means window closed. Do this here:
```windowEvent.consume()```
By doing this you do not need `windowEvent` variable in class level. remove it.

File: src/main/java/ir/ac/sbu/controller/MainController.java [hunk 19]
```diff
@@ -201,6 +207,8 @@ protected void updateItem(GraphModel item, boolean empty) {
         helpLicenseMenuItem.setOnAction(this::helpLicense);
         helpManualMenuItem.setOnAction(this::helpManual);
         addGraphBtn.setOnAction(this::addGraph);
+
+        setOnCloseRequest();
```
Comment (Borjianamin98) [inline src/main/java/ir/ac/sbu/controller/MainController.java]: Do this I think better:
```
Main.getPrimaryStage().setOnCloseRequest(this::setOnCloseRequest);
...
private void setOnCloseRequest(WindowEvent event) {
...
}
```

**Review by Borjianamin98 — CHANGES_REQUESTED** [review]

File: src/main/java/ir/ac/sbu/controller/MainController.java (hunk 17 above)
Comment (hamidhandid) [reply src/main/java/ir/ac/sbu/controller/MainController.java]: No, we should consume the event when user select yes in save dialog but does not select a path for file (select cancel in save dialog). So, PGen should consume the event and not close the program.

Comment (hamidhandid) [reply src/main/java/ir/ac/sbu/controller/MainController.java]: My Suggestion is:

```java
private File fileSave(ActionEvent actionEvent) {
        if (lastChosenFileForSave != null) {
            saveGraphs(lastChosenFileForSave);
            return lastChosenFileForSave;
        } else {
            return fileSaveAs(actionEvent);
        }
    }
```

```java
private File fileSaveAs(ActionEvent actionEvent) {
        File selectedFile = DialogUtility.showSaveDialog(pane.getScene().getWindow(), "Save PGen File", "*.pgs");
        if (selectedFile != null) {
            lastChosenFileForSave = selectedFile;
            saveGraphs(lastChosenFileForSave);
        }
        return selectedFile;
    }
```

```java
private void setOnCloseRequest(WindowEvent event) {
        String dialogString = lastChosenFileForSave != null ? "Do you want to save changes?" : "Do you want to save changes to a file?";
        Optional<ButtonType> result = DialogUtility.showConfirmationDialog("PGEN", dialogString);
        if (!result.isPresent() || result.get().equals(ButtonType.CANCEL)) {
            event.consume();
            return;
        }
        ButtonType selectedButton = result.get();
        if (selectedButton.equals(ButtonType.YES)) {
            File file = fileSave(null);
            if (file == null) {
                event.consume();
                return;
            }
        }
        Platform.exit();
        System.exit(0);
    }
}
```

Comment (hamidhandid) [reply src/main/java/ir/ac/sbu/controller/MainController.java]: Last function can get better like this:

```java
private void setOnCloseRequest(WindowEvent event) {
        String dialogString = lastChosenFileForSave != null ? "Do you want to save changes?" : "Do you want to save changes to a file?";
        Optional<ButtonType> result = DialogUtility.showConfirmationDialog("PGEN", dialogString);
        if (!result.isPresent() || result.get().equals(ButtonType.CANCEL) || (result.get().equals(ButtonType.YES) && fileSave(null) == null)) {
            event.consume();
            return;
        }
        Platform.exit();
        System.exit(0);
    }
```

Comment (Borjianamin98) [reply src/main/java/ir/ac/sbu/controller/MainController.java]: Add this comment for "if conditions" in "setOnCloseRequest" method:

> Although we have considered the user to cancel after the dialog is displayed, the user may select the "YES" option at first, but then close the opened window without selecting any path or without pressing any buttons. In this case, the person has practically given up and we do not want to close the program, so we check whether the user has actually saved the file or not. 

Also return "boolean" for save methods which show whether we actually saved file or not. because providing a last saved file is not useful and also it is in class variables. (It is better to show that file saved or not)

**Comment by Borjianamin98** [comment]
Also increase version of application.

File: src/main/java/ir/ac/sbu/controller/MainController.java (hunk 17 above)
Comment (hamidhandid) [reply src/main/java/ir/ac/sbu/controller/MainController.java]: Edit:

```java
private boolean fileSave(ActionEvent actionEvent) {
        if (lastChosenFileForSave != null) {
            saveGraphs(lastChosenFileForSave);
            return true;
        } else {
            return fileSaveAs(actionEvent);
        }
    }

    private boolean fileSaveAs(ActionEvent actionEvent) {
        File selectedFile = DialogUtility.showSaveDialog(pane.getScene().getWindow(), "Save PGen File", "*.pgs");
        if (selectedFile != null) {
            lastChosenFileForSave = selectedFile;
            saveGraphs(lastChosenFileForSave);
            return true;
        }
        return false;
    }

private void setOnCloseRequest(WindowEvent event) {
        String dialogString = lastChosenFileForSave != null ? "Do you want to save changes?" : "Do you want to save changes to a file?";
        Optional<ButtonType> result = DialogUtility.showConfirmationDialog("PGEN", dialogString);
        if (!result.isPresent() || result.get().equals(ButtonType.CANCEL)
                // Although we have considered the user to cancel after the dialog is displayed,
                // the user may select the "YES" option at first,
                // but then close the opened window without selecting any path or without pressing any buttons.
                // In this case, the person has practically given up and we do not want to close the program,
                // so we check whether the user has actually saved the file or not.
                || (result.get().equals(ButtonType.YES) && fileSave(null))) {
            event.consume();
            return;
        }
        Platform.exit();
        System.exit(0);
    }
```

Comment (Borjianamin98) [reply src/main/java/ir/ac/sbu/controller/MainController.java]: Excellent.
